### PR TITLE
[Split] random split method improvement

### PR DESCRIPTION
**Pull request by XinweiHe**
For https://github.com/pyg-team/pytorch-frame/issues/352

### CODE REVIEW

File: test/utils/test_split.py [hunk 1]
```diff
@@ -9,13 +11,26 @@ def test_generate_random_split():
...
     test_ratio = 0.1
 
-    split = generate_random_split(num_data, seed=42, train_ratio=train_ratio,
-                                  val_ratio=val_ratio)
-    assert (split == SPLIT_TO_NUM['train']).sum() == int(num_data *
-                                                         train_ratio)
-    assert (split == SPLIT_TO_NUM['val']).sum() == int(num_data * val_ratio)
-    assert (split == SPLIT_TO_NUM['test']).sum() == int(num_data * test_ratio)
+    split = generate_random_split(num_data, seed=42,
+                                  ratios=[train_ratio, val_ratio])
```
Comment (weihua916) [inline test/utils/test_split.py:15]: why do we want this to be a list? I think the previous argument is clearer.

File: torch_frame/data/dataset.py [hunk 2]
```diff
@@ -707,20 +708,25 @@ def get_split(self, split: str) -> Dataset:
...
+        val_set = self.get_split(TrainingStage.VAL)
+        test_set = self.get_split(TrainingStage.TEST)
+        if test_set.num_rows == 0:
+            test_set = None
+        return train_set, val_set, test_set
+
+    def random_split(self, ratios: list[float] | None = None):
+        split = generate_random_split(self.num_rows, ratios)
+        self.split_col = 'split'
+        self.df[self.split_col] = split
```
Comment (weihua916) [inline torch_frame/data/dataset.py:729]: I think it'd be nice to supply an argument to control the random seed of the random seed.

Comment (weihua916) [reply torch_frame/data/dataset.py:729]: Also, add doc-string so that people know how to use it.

File: torch_frame/typing.py [hunk 3]
```diff
@@ -28,6 +28,12 @@ def supports_task_type(self, task_type: 'TaskType') -> bool:
         return self in task_type.supported_metrics
 
 
+class TrainingStage(Enum):
```
Comment (weihua916) [inline torch_frame/typing.py:31]: Brief doc-string.

File: torch_frame/utils/split.py [hunk 4]
```diff
@@ -1,30 +1,84 @@
...
+        raise ValueError("No more than three training splits is supported")
+    if len(ratio) < 1:
+        raise ValueError("At least two training splits are required")
+
+    for val in ratio:
+        if val < 0:
+            raise ValueError("'ratio' can not contain negative values")
+
+    if sum(ratio) - 1 > 1e-2:
+        raise ValueError("'ratio' exceeds more than 100% of the data")
```
Comment (weihua916) [inline torch_frame/utils/split.py:84]: Could you clarify what's wrong with the old implementation? I think the old one is simpler.

File: test/utils/test_split.py (hunk 1 above)
Comment (XinweiHe) [reply test/utils/test_split.py:15]: This format is more aligned with split method in torch https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#random_split. And that to support split with only train and validation data with the new format, we are able to do `split(ratios=[0.8], ...)` while with the old format, we could only do `split(train_ratio=0.8, val_ratio=None, ...)` which is a bit weird, because this indicates a train/test split instead of a train/val split which is more common.

Comment (weihua916) [reply test/utils/test_split.py:15]: Yeah but here we are actually assigning TRAIN to the first ratio, VAL to the second ratio, and so on. So I still think the previous one is better.

Comment (weihua916) [reply test/utils/test_split.py:15]: torch's random split makes sense since it just splits the dataset. There is no train/val/test assignment happening under the hood.

File: torch_frame/data/dataset.py [hunk 5]
```diff
@@ -695,7 +696,7 @@ def col_select(self, cols: ColumnSelectType) -> Dataset:
 
         return dataset
 
-    def get_split(self, split: str) -> Dataset:
+    def get_split(self, split: TrainingStage) -> Dataset:
```
Comment (weihua916) [inline torch_frame/data/dataset.py:699]: I think it's nicer to accept `str` as input here. It's sometimes troublesome to import `TrainingStage`.

**Comment by weihua916** [comment]
Thanks for the effort. I am not sure the additional complexity is worthwhile though. I think the generalization to `ratios: list[float]` is not necessary here since each element has its role of train/val/test ratios.

File: torch_frame/data/dataset.py (hunk 5 above)
Comment (XinweiHe) [reply torch_frame/data/dataset.py:699]: Maybe we can support both? i.e. `class TrainingStage(str, Enum)`  I feel like `TrainingStage` has its advantage in that no mistypes will occur that causes unexpected behavior, e.g. 'Train' vs 'train'. I feel like this is also the reason we support lots of Enum classes in the codebase

Comment (weihua916) [reply torch_frame/data/dataset.py:699]: Yes, we can support both.

Comment (weihua916) [reply torch_frame/data/dataset.py:699]: `Union[str, TrainingStage]`

File: test/utils/test_split.py (hunk 1 above)
Comment (XinweiHe) [reply test/utils/test_split.py:15]: In the previous format we always assume we want to have a test set. So
1. it does not have the flexibility to doing something like (train: 0.3, val: 0.1, test: 0.2)
2. when we do train: 0.8, val: 0.0. Then the inferred test ratio is 0.2. Which is a bit weird because in this case only train and val (test) split is well-defined. i.e. we only intend to split the data into 2 parts instead of 3 parts. But we end up with an additional validation set that is completely empty

We could also keep the previous format but support the case where val ratio could be `None` which indicates there is no validation set instead of an empty validation set if you have strong opinion here.

Comment (weihua916) [reply test/utils/test_split.py:15]: I understand you. but I still want features based on needs.
I don't find these logics useful for now, and having new interface would break other parts. For instance, this PR does not modify `generate_random_split` used in `data_frame_benchmark.py` and elsewhere.

Comment (XinweiHe) [reply test/utils/test_split.py:15]: > For instance, this PR does not modify generate_random_split used in data_frame_benchmark.py and elsewhere

This is still a draft now. I kind of want to sync with you on this first before changing other parts of the code
